docs/fileformat-prose.rst

```diff
@@ -100,6 +100,11 @@ indptr <HDF5 dataset "indptr": shape (38411,), type "<i4">
 ..    indices                  Dataset {41459314/Inf}
 ..    indptr                   Dataset {38411/Inf}
 
+AwkwardArrays
+~~~~~~~~~~~~~
+
+TODO
```

Comment:
@ivirshup, marking this to be resolved

Reply:
````suggestion
Ragged arrays are supported in `anndata` through the [Awkward Array](https://awkward-array.org/) library. For storage on disk, we break down the awkward array into it's constituent arrays using [`ak.to_buffers`](https://awkward-array.readthedocs.io/en/latest/_auto/ak.to_buffers.html), then writing these arrays using `anndata`'s methods.

The container of arrays is stored in a group called `"container"`

```python
>>> import zarr
>>> z = zarr.open("airr.zarr", "r")
>>> awkward_group = z["obsm/airr"]
>>> awkward_group.tree()
```

```
airr
 └── container
     ├── node0-offsets (17,) int64
     ├── node2-offsets (40,) int64
     ├── node3-data (117,) uint8
     ├── node4-offsets (40,) int64
     └── node5-data (117,) uint8
```

The length of the array is saved to it's own `"length"` attribute, while metadata for the array structure is serialized and saved to the "form" attribute.

```python
>>> dict(awkward_group.attrs)
```

```json
{'encoding-type': 'awkward-array',
 'encoding-version': '0.1.0',
 'form': '{"class": "ListOffsetArray", "offsets": "i64", "content": {"class": '
         '"RecordArray", "contents": {"locus": {"class": "ListOffsetArray", '
         '"offsets": "i64", "content": {"class": "NumpyArray", "primitive": '
         '"uint8", "inner_shape": [], "has_identifier": false, "parameters": '
         '{"__array__": "char"}, "form_key": "node3"}, "has_identifier": '
         'false, "parameters": {"__array__": "string"}, "form_key": "node2"}, '
         '"junction_aa": {"class": "ListOffsetArray", "offsets": "i64", '
         '"content": {"class": "NumpyArray", "primitive": "uint8", '
         '"inner_shape": [], "has_identifier": false, "parameters": '
         '{"__array__": "char"}, "form_key": "node5"}, "has_identifier": '
         'false, "parameters": {"__array__": "string"}, "form_key": "node4"}}, '
         '"has_identifier": false, "parameters": {}, "form_key": "node1"}, '
         '"has_identifier": false, "parameters": {}, "form_key": "node0"}',
 'length': 16}
```

These can be read back as awkward arrays using the [`ak.from_buffers`](https://awkward-array.readthedocs.io/en/latest/_auto/ak.from_buffers.html) function:

```python
>>> import awkward._v2 as ak
>>> from anndata.experimental import read_elem
>>> ak.from_buffers(
...     awkward_group.attrs["form"],
...     awkward_group.attrs["length"],
...     {k: read_elem(v) for k, v in awkward_group["container"].items()}
... )
```

```
<Array [[], [...], ..., [{locus: 'TRD', ...}]] type='16 * var * {locus: str...'>
```

````

Reply:
^ My suggestion. I'm wondering if we should get rid of the extra `container` group.

Reply:
looks good to me, but shouldn't this be in `rst` rather than markdown? Or do the anndata docs already support MyST? 

I don't have an opinion on the `container` -- I know to little about h5 and zarr files. 

Is the `airr.zarr` folder supposed to be available somewhere or is it just an arbitrary example?

Reply:
> Is the airr.zarr folder supposed to be available somewhere or is it just an arbitrary example?

Arbitrary example, as with the other examples here.

> looks good to me, but shouldn't this be in rst rather than markdown? Or do the anndata docs already support MyST?

Oh! Thought I'd already done that. I've edited the suggestion to rst.